src/Microsoft.AspNetCore.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs

```diff
@@ -116,6 +152,11 @@ internal static void AddRazorViewEngineServices(IServiceCollection services)
                 return new DefaultChunkTreeCache(accessor.FileProvider);
             }));
 
+            services.TryAddTransient<ITagHelperDescriptorResolver, TagHelperDescriptorResolver>();
+
+            services.TryAddTransient<ITagHelperTypeResolver, TagHelperTypeResolver>();
+            services.TryAddTransient<ITagHelperDescriptorFactory>(s => new TagHelperDescriptorFactory(designTime: false));
```

Comment:
We usually only make top level objects as singletons as the container has to hold on to them